pkg/v1/cli/carvelhelpers/registry.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2021 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package carvelhelpers
+
+import (
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	ctlimg "github.com/k14s/imgpkg/pkg/imgpkg/registry"
+	"github.com/pkg/errors"
+
+	"github.com/vmware-tanzu/tanzu-framework/pkg/v1/tkg/constants"
+	"github.com/vmware-tanzu/tanzu-framework/pkg/v1/tkg/tkgconfigbom"
+	"github.com/vmware-tanzu/tanzu-framework/pkg/v1/tkg/tkgconfigpaths"
+	"github.com/vmware-tanzu/tanzu-framework/pkg/v1/tkg/utils"
+	"github.com/vmware-tanzu/tanzu-framework/pkg/v1/tkr/pkg/registry"
+)
+
+// GetFilesMapFromImage returns map of files metadata
+// It takes os environment variables for custom repository and proxy
+// configuration into account while downloading image from repository
+func GetFilesMapFromImage(imageWithTag string) (map[string][]byte, error) {
+	reg, err := newRegistry()
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to initialize registry")
+	}
+	return reg.GetFiles(imageWithTag)
+}
+
+// ReadImageAndSaveFilesToTempDir reads OCI image and saves file to temp dir
+// returns temp configuration dir with downloaded imgpkg bundle
+func ReadImageAndSaveFilesToTempDir(imageWithTag string) (string, error) {
+	filesMap, err := GetFilesMapFromImage(imageWithTag)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get resource files from discovery")
+	}
+
+	tmpDir, err := os.MkdirTemp("", "oci_image")
+	if err != nil {
+		return "", errors.Wrap(err, "error creating temporary directory")
+	}
+
+	for filename, fileData := range filesMap {
+		err = utils.SaveFile(filepath.Join(tmpDir, filename), fileData)
+		if err != nil {
+			return "", errors.Wrap(err, "error while saving file")
+		}
+	}
+	return tmpDir, nil
+}
+
+// newRegistry returns new
+func newRegistry() (registry.Registry, error) {
+	verifyCerts := true
+	skipVerifyCerts := os.Getenv(constants.ConfigVariableCustomImageRepositorySkipTLSVerify)
+	if strings.EqualFold(skipVerifyCerts, "true") {
+		verifyCerts = false
+	}
+
+	registryOpts := &ctlimg.Opts{
+		VerifyCerts: verifyCerts,
+		Anon:        true,
+	}
+
+	if runtime.GOOS == "windows" {
+		err := tkgconfigbom.AddRegistryTrustedRootCertsFileForWindows(registryOpts)
```

Comment:
nit: It feels odd to be calling into tkgconfigbom to setup the Registry object. But if we do, it would be more DRY to facto r out the functionality of InitBomRegistry common with this code into a different function, and have newRegistry call said function instead.

Reply:
Thanks. Refactored the code and moved it to `clientconfighelpers` pkg.